coding-questions-python/climbing-stairs.py

```diff
@@ -0,0 +1,57 @@
+'''
+Approach (recursion):
+    - at each step, look at all combinations, i.e. 1 and 2
+    - size of the recursion tree will be 2^n
+    - so time complexity will be O(2^n)
+    - space complexity is O(n) as the dept of recursion can go up to n
+
+Approach (recursion + memoization):
+    - same approach as above
+    - along with maintaining a dict to store + reuse results
+    - size of the recursion tree can go upto n #TODO: understand this
+    - so time complexity will be O(n)
+    - space complexity is also O(n) as the dept of recursion can go up to n
+
+Dynamic Programming Approach:
+    - this problem contains the optimal substructure property
+    (optimal solution can be constructed from optimal sols of it's subproblems)
+    - one can reach ith step by doing:
+        * taking a step step from (i-1)th step
+        * taking two steps from (i-2)th step
+    - so, total number of ways to reach ith step is the sum of ways to reach
+        (i-1)th step and ways to reach (i-2)th step
+    - if step[i] denotes the number of ways to reach the ith step, then
+        step[i] = step[i-1] + step[i-2]
+    - TODO: time and space complexity
```

Comment:
Verify that the time and space complexities for all the three approaches are correct and make sense